Mathlib/Data/QuotType.lean

```diff
@@ -0,0 +1,564 @@
+/-
+Copyright (c) 2024 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+module
+
+public import Mathlib.Logic.Relator
+public import Mathlib.Logic.Unique
+public meta import Qq.MetaM
+
+/-!
+# Typeclass for quotient types
+
+This typeclass is primarily for use by type synonyms of `Quot` and `Quotient`.
+
+Using `QuotType` API for type synonyms of `Quot` and `Quotient` will avoid defeq abuse caused by
+directly using `Quot` and `Quotient` APIs.
+
+## Main definitions
+
+* `QuotType Q α r` : the type `Q` is canonically isomorphic to `Quot α r`.
+                     used for deriving `QuotType` instances from the output type.
+* `QuotTypeOut`    : used for deriving `QuotType` instances from the input type.
+* `mkQ`            : the quotient map inferred from the output type
+
+## Notations
+
+* `⟦a⟧`            : the quotient map inferred from the output type
+* `mkQ'` `⟦a⟧'`    : the quotient map inferred from the input type
+                       via typeclass `QuotTypeOut`
+-/
+
+@[expose] public section
+
+/--
+The class `QuotType Q α r` expresses the type `Q` is canonically isomorphic to `Quot α r`.
+
+Using `QuotType` API for type synonyms of `Quot` and `Quotient` will avoid defeq abuse of the types
+caused by directly using `Quot` and `Quotient` APIs.
+-/
+class QuotType (Q : Sort*) (α : outParam Sort*) (r : outParam (α → α → Prop)) where
+  /-- The quotient map. Inferred from the output type. -/
+  mkQ : α → Q := by exact Quot.mk _
+  /-- The map from the given quotient type to `Quot r`. -/
+  toQuot : Q → Quot r := by exact (·)
+  toQuot_mkQ : ∀ a, toQuot (mkQ a) = Quot.mk r a := by exact fun _ ↦ rfl
+  mkQ_surjective : Function.Surjective mkQ := by exact Quot.exists_rep
+  /--
+  The analogue of `Quot.sound`: If `a` and `b` are related by the relation,
+  then they have equal equivalence classes.
+  -/
+  sound {a b : α} : r a b → mkQ a = mkQ b := by exact Quot.sound
+
+export QuotType (mkQ)
+
+/--
+`QuotTypeOut` is used for deriving `QuotType` instances from the input type.
+
+The instances of `QuotTypeOut` may be defined for types that have specific uses.
+
+```
+instance : QuotTypeOut ZFSet PSet PSet.Equiv where
+```
+
+They are also usually defined as scoped or local instances.
+
+```
+scoped instance {α} [s : Setoid α] : QuotTypeOut (Quotient s) α (· ≈ ·) where
+```
+-/
+class QuotTypeOut (Q : outParam Sort*) (α : Sort*) (r : outParam (α → α → Prop))
+    [QuotType Q α r] : Prop where
+
+namespace QuotType
+
+attribute [simp] toQuot_mkQ
+
+open Lean Elab Term Meta Qq
+
+@[inherit_doc mkQ]
+elab "⟦" a:term "⟧" : term <= Q => do
```

Comment:
When attempting to replace existing mathlib APIs with `QuotType`, I discovered that Lean attempts to synthesize `QuotType (SomeQuotientOf α) (Subtype (_ : α → Prop)) ?_`. The original API could automatically insert the coercion.

I don't know if this can be considered a core issue. (AFAIK core is unlikely to change features related to typeclasses now) I'm also not sure if this issue is worth resolving. But I made this elaborator to ensure that the quotient map is derived from the output type.